### DISCUSSION TITLE
Exiting umtprd on Linux Hangs

### DISCUSSION THREAD

**Discussion by kstarfss** [viveris/uMTP-Responder#63] [Q&A]
Hello uMTP Responder group,
I am new to MTP Responder and have been working on a project that uses it and my colleagues have it all configured and working using FunctionFS.  We are using version 1.3.10 and everything is working good except when we try to shut it down.  We are running on embedded Linux and when we try to reboot, shutdown or press power button which invokes a shutdown, the system hangs.  We see TERM  and KILL signals sent but the system hangs at that point.  I can't even send a kill -1, or -15, or -9 myself as root to kill it without it hanging.   Without **umtprd** running, the system shuts down and powers off with no problems.  We traced it down to what looks like a Linux Kernel USB driver issue in the **f_fs** module.  We are using Linux Kernel 5.4.0 and I have applied numerous patches from kernel.org that were related to this area and none of them have fixed the problem.  So, this seems to be a Linux Kernel issue in the USB gadget driver f_fs area and not an MTP Responder issue specifically which is why I am asking in this discussion forum if anyone else has had this problem while trying to exit uMTP Responder and resolved it?  Do I just need to apply the right Kernel USB patch to correct the issue and if so does anyone know which one(s)?  I don't think this is unique to our device and others have seen this issue too but I haven't found a solution short of debugging the kernel usb driver myself.  As stated, everything is working except when we try to power down our device.  Any help or clues from anyone is appreciated.

**Comment by jfdelnero** [comment]
I think that a message queue or an signal handler is probably needed to close the usb pipes and handlers in the right order.

**Reply by kstarfss** [reply]
Thank you for replying and suggestions.  I added what you suggested to close the usb pipes and handlers to my signal handler and tried some other things as well in the signal handler.  I can see the signal handler get invoked and close everything successfully and exit, but the OS still locks up. It is something after this point in the kernel.  I have been able to work around it in our power handler driver when the power button is pressed but when doing a soft shutdown on the console it takes a different path and still locks up.  I will debug further and work with Linux kernel forum to try and resolve.  Thank you for your help.

**Reply by mitchmandler** [reply]
Were you able to figure out how to get around the hang?  I am seeing the same issue and tried adding a signal handler.  But no matter what I do in the signal handler, the io_thread() remains blocked in read_usb() call and prevents umtprd from exiting
